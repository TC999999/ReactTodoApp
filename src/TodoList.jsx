import Todo from "./Todo.jsx";
import NewTodoForm from "./NewTodoForm.jsx";
import EditTodoForm from "./EditTodoForm.jsx";
import { useState } from "react";
import { v4 as uuid } from "uuid";

const TodoList = () => {
  const initialState = () => {
    if (!localStorage.getItem("todos")) {
      return [];
    } else {
      return JSON.parse(localStorage.getItem("todos"));
    }
  };
  const [todos, setTodos] = useState(initialState());
  const [needsEditing, setNeedsEditing] = useState(false);
  const [editedTodo, setEditedTodo] = useState({ id: "", todo: "" });

  const addTodo = (newTodo) => {
    const newTodos = [...todos, { ...newTodo, id: uuid(), complete: false }];
    setTodos((todos) => newTodos);
    localStorage.setItem("todos", JSON.stringify(newTodos));
  };

  const delTodo = (e) => {
    let todosCopy = todos.map((t) => {
      return { ...t };
    });

    let id = e.target.closest(".todo-item").id;
    let filteredTodosCopy = todosCopy.filter((t) => {
      return t.id !== id;
    });
    setNeedsEditing(false);
    setTodos(filteredTodosCopy);
    localStorage.setItem("todos", JSON.stringify(filteredTodosCopy));
  };

  const showEditForm = (e) => {
    if (needsEditing == false) {
      let id = e.target.closest(".todo-item").id;

      const target = todos.filter((t) => {
        return t.id == id;
      });
      setNeedsEditing(true);
      setEditedTodo({ id: id, todo: target[0].todo });
    } else {
      setEditedTodo({ id: "", todo: "" });
      setNeedsEditing(false);
    }
  };

  const editTodo = (newTodo) => {
    const todosCopy = todos.map((t) => {
      if (editedTodo.id == t.id) {
        return { ...t, todo: newTodo["edit-todo"] };
      } else {
        return { ...t };
      }
    });
    setNeedsEditing(false);
    setEditedTodo({ id: "", todo: "" });
    setTodos(todosCopy);
    localStorage.setItem("todos", JSON.stringify(todosCopy));
  };

  const compTodo = (e) => {
    let task = e.target.previousElementSibling;
    let id = e.target.closest(".todo-item").id;
    let todosCopy;

    if (!task.classList.contains("complete")) {
      todosCopy = todos.map((t) => {
        if (t.id == id) {
          return { ...t, complete: true };
        } else {
          return { ...t };
        }
      });
    } else {
      todosCopy = todos.map((t) => {
        if (t.id == id) {
          return { ...t, complete: false };
        } else {
          return { ...t };
        }
      });
    }
    setTodos(todosCopy);
    localStorage.setItem("todos", JSON.stringify(todosCopy));
  };

  return (
    <div id="todos">
      <h3>To-Do List</h3>
      {!needsEditing && <NewTodoForm addTodo={addTodo} />}

      {needsEditing && (
        <EditTodoForm
          todo={editedTodo.todo}
          editTodo={editTodo}
          showEditForm={showEditForm}
        />
      )}
      {!needsEditing && (
        <div id="todo-list-div">
          {todos.map(({ id, todo, complete }) => {
            return (
              <Todo
                id={id}
                key={id}
                todo={todo}
                complete={complete}
                delTodo={delTodo}
                compTodo={compTodo}
                showEditForm={showEditForm}
              />
            );
          })}
        </div>
      )}
    </div>
  );
};

export default TodoList;
