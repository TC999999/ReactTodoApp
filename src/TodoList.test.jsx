import { fireEvent, render, screen } from "@testing-library/react";
import TodoList from "./TodoList.jsx";

it("renders without crashing", function () {
  render(<TodoList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should make a task when the form is submitted", function () {
  const { queryByText, getByLabelText } = render(<TodoList />);
  const input = getByLabelText("Next on the List");
  const btn = queryByText("Add Todo");
  expect(queryByText("- test-task")).not.toBeInTheDocument();

  fireEvent.change(input, { target: { value: "test-task" } });
  fireEvent.submit(btn);
  expect(queryByText("- test-task")).toBeInTheDocument();
});

it("should add a 'complete' class to a task when 'Mark As Complete' button is clicked", function () {
  const { queryByText } = render(<TodoList />);
  const task = queryByText("- test-task");
  expect(task.className).toBe("task ");
  const mark = queryByText("Mark As Complete");
  fireEvent.click(mark);
  expect(task.className).toBe("task complete");
  expect(queryByText("Mark As Complete")).not.toBeInTheDocument();
});

it("should remove the 'complete' class when 'Mark as Incomplete' button is clicked", function () {
  const { queryByText } = render(<TodoList />);
  const task = queryByText("- test-task");
  expect(task.className).toBe("task complete");
  const mark = queryByText("Mark As Incomplete");
  fireEvent.click(mark);
  expect(task.className).toBe("task ");
  expect(queryByText("Mark As Incomplete")).not.toBeInTheDocument();
});

it("should hide the todo-list and new todo form and show edit form when 'Edit Task' button is clicked", function () {
  const { queryByText, container } = render(<TodoList />);

  expect(
    container.querySelector("div[id='todo-list-div']")
  ).toBeInTheDocument();
  expect(
    container.querySelector("div[id='todo-form-div']")
  ).toBeInTheDocument();
  expect(
    container.querySelector("div[id='edit-todo-div']")
  ).not.toBeInTheDocument();

  const edit = queryByText("Edit Task");
  fireEvent.click(edit);

  expect(
    container.querySelector("div[id='todo-list-div']")
  ).not.toBeInTheDocument();
  expect(
    container.querySelector("div[id='todo-form-div']")
  ).not.toBeInTheDocument();
  expect(
    container.querySelector("div[id='edit-todo-div']")
  ).toBeInTheDocument();
});

it("should cancel any changes if the 'Cancel' button is clicked when editing a function", function () {
  const { queryByText, getByLabelText, container } = render(<TodoList />);

  expect(queryByText("- test-task")).toBeInTheDocument();
  expect(queryByText("- new-test-task")).not.toBeInTheDocument();

  const edit = queryByText("Edit Task");
  fireEvent.click(edit);
  const input = getByLabelText("Edit Task - test-task");
  fireEvent.change(input, { target: { value: "new-test-task" } });
  expect(input).toHaveValue("new-test-task");
  fireEvent.click(queryByText("Cancel"));

  expect(queryByText("- test-task")).toBeInTheDocument();
  expect(queryByText("- new-test-task")).not.toBeInTheDocument();
});

it("should hide the edit form, show the todo-list with the changed task when the edit button is clicked", function () {
  const { queryByText, getByLabelText, container } = render(<TodoList />);

  expect(queryByText("- test-task")).toBeInTheDocument();
  expect(queryByText("- new-test-task")).not.toBeInTheDocument();

  const edit = queryByText("Edit Task");
  fireEvent.click(edit);
  const input = getByLabelText("Edit Task - test-task");
  fireEvent.change(input, { target: { value: "new-test-task" } });
  fireEvent.submit(queryByText("Edit Todo"));

  expect(queryByText("- test-task")).not.toBeInTheDocument();
  expect(queryByText("- new-test-task")).toBeInTheDocument();

  expect(
    container.querySelector("div[id='todo-list-div']")
  ).toBeInTheDocument();
  expect(
    container.querySelector("div[id='todo-form-div']")
  ).toBeInTheDocument();
  expect(
    container.querySelector("div[id='edit-todo-div']")
  ).not.toBeInTheDocument();
});

it("should delete a task when the 'X' button is clicked", function () {
  const { queryByText } = render(<TodoList />);

  expect(queryByText("- new-test-task")).toBeInTheDocument();
  const del = queryByText("X");

  fireEvent.click(del);
  expect(queryByText("- new-test-task")).not.toBeInTheDocument();
});
