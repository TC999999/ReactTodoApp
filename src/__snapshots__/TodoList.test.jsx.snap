// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    id="todos"
  >
    <h3>
      To-Do List
    </h3>
    <div
      id="todo-form-div"
    >
      <form
        id="todo-form"
      >
        <label
          for="todo"
        >
          Next on the List
        </label>
        <input
          id="todo"
          name="todo"
          placeholder="Enter Task Here"
          type="text"
          value=""
        />
        <button>
          Add Todo
        </button>
      </form>
    </div>
    <div
      id="todo-list-div"
    />
  </div>
</DocumentFragment>
`;
